### PR TITLE
Add ReLU to ATen

### PR DESCRIPTION
**Pull request by goldsborough**
Adds relu to ATen. There's `clamp_min()`, but `relu()` makes sense given that we have `selu()`, `prelu()`, `rrelu()` etc. (working through https://github.com/pytorch/pytorch/issues/4619)

Fixes https://github.com/pytorch/pytorch/issues/3847

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -308,6 +308,10 @@
 
 - func: select(Tensor self, int64_t dim, int64_t index) -> Tensor
 
+- func: relu(Tensor self) -> Tensor
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: Should this also be function-only?

Comment (SsnL) [reply aten/src/ATen/native/native_functions.yaml]: Yeah it will generate both. I just mentioned because I saw rrelu and selu are function-only. Maybe @colesbury has more to say on this.

**Comment by apaszke** [comment]
Just curious what is our strategy regarding nn functions in ATen. They're not part of the `torch` namespace (for a good reason), and I think it would be clearer to have a `at::nn` namespace (or even `using F = at::nn::functional`) where they would live. Is the plan to just temporarily add them to the main namespace and only move them later? Or do we want `at::` to expose this huuuuuuuuge set of functions?

**Comment by gchanan** [comment]
I'm not sure we have a plan, and I agree organization could be useful here, but why draw the barrier at `nn` vs non-`nn`, outside of THNN historical reasons?  Do people use pytorch for non-nn use cases?

**Comment by SsnL** [comment]
Yeah, people use it for more than deep learning  purposes, e.g. PPL (Pyro),
HamIltonian Monte Carlo, or just CUDA-accelerated NumPy.

**Comment by gchanan** [comment]
What's the subset of functionality that each of those purposes use?

**Comment by SsnL** [comment]
It can vary, but they mostly use random sampling, linear algebra methods,
autograd (mostly through linalg methods rather than through NN layers), and
sometimes gradient-based optimizers.

**Review by colesbury — COMMENTED** [review]
I think you should update the implementation in torch/nn/functional.py (probably use the `_add_docstr` pattern).

That implementation uses `threshold`. It should have the same semantics, but there may be performance differences (including related to how the derivative is implemented.)

Since `relu` is used often, it's probably worth comparing them or just call `threshold` to be safe.

**Comment by goldsborough** [comment]
So forwards seems faster, but backwards is much slower? How? Any ideas?

```py
In [1]: import torch
In [2]: x = torch.randn(100, 100, requires_grad=True)
In [3]: %timeit -n1000 torch.relu(x).sum().backward()
213 µs ± 15.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
In [4]: %timeit -n1000 torch.nn.functional.threshold(x, 0, 0).sum().backward()
118 µs ± 12.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
In [5]: %timeit -n1000 torch.relu(x)
11.7 µs ± 107 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
In [6]: %timeit -n1000 torch.nn.functional.threshold(x, 0, 0)
14.4 µs ± 951 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by zou3519** [comment]
@goldsborough Intermediate tensors are generated as part of clamp_min's backwards:
image
(In derivatives.yaml).

On the other hand, the current `relu` calls `threshold` which is in THNN (THNN/generic/Threshold.c). It has a custom backward function that does not generate intermediate tensors.

**Comment by goldsborough** [comment]
Ok thanks for the insight @zou3519  -- I'll try your suggestion

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -716,6 +716,9 @@
 - name: soft_margin_loss_forward(Tensor self, Tensor target, bool size_average, bool reduce)
   self: soft_margin_loss_backward(grad, self, target, size_average, reduce)
 
+- name: relu(Tensor self)
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml]: nit: You don't need to specify a derivative for `relu` if you define `relu` with `threshold`:
```
Tensor relu(const Tensor & self) {
  return self.threshold(0, 0);
}
```
but then it'll look different from the definition of `relu_` (there isn't a `threshold_` function I think)

Comment (goldsborough) [reply tools/autograd/derivatives.yaml]: It seems that `relu` with `clamp_min` is slightly faster than `threshold` too though?

Comment (zou3519) [reply tools/autograd/derivatives.yaml]: Good point. That's interesting:

image
(threshold on left, clamp_min on right (as cmaxValue))

I don't think the conditional should slow down the code by as much as you're seeing but... that can be an investigation for another day

**Comment by goldsborough** [comment]
@colesbury @apaszke any more thoughts? I think there's a larger discussion to be had about separating NN functionality from Tensor computation functionality within ATen, but this PR just backfills the missing C++ version of `relu`, so hopefully we can defer that discussion.

**Comment by apaszke** [comment]
I am a bit ambivalent about this PR. This is not really a pressing issue for C++ users, because they can define a ReLU alias as a one liner. What's more, the refactor that would separate nn parts from general linear algebra parts would also cause breakages for them so it's even safer for them to do it that way.

Additionally, this adds `Tensor.relu` that we never had, and never plan to have. If we **really** want to merge this, then I would be ok with having it as a namespace only thing in C++ (left for a clean up later), but it *should never appear in the `torch.` namespace in Python* (nor as a Variable method).

**Review by apaszke — APPROVED** [review]
